File: backend/internal/database/models/bulk_operations.go
package models

import (
	"time"

	"github.com/google/uuid"
)

// BulkImportResult represents the result of a bulk guest import operation
type BulkImportResult struct {
	TotalProcessed int64               `json:"total_processed"`
	SuccessCount   int64               `json:"success_count"`
	FailureCount   int64               `json:"failure_count"`
	Errors         []BulkImportError   `json:"errors"`
	Warnings       []BulkImportWarning `json:"warnings"`
	NewGuests      []Guest             `json:"new_guests"`
	UpdatedGuests  []Guest             `json:"updated_guests"`
	ImportTime     time.Duration       `json:"import_time"`
}

// BulkImportError represents an error during bulk import
type BulkImportError struct {
	Row     int    `json:"row"`
	Field   string `json:"field"`
	Value   string `json:"value"`
	Message string `json:"message"`
}

// BulkImportWarning represents a warning during bulk import
type BulkImportWarning struct {
	Row     int    `json:"row"`
	Field   string `json:"field"`
	Value   string `json:"value"`
	Message string `json:"message"`
}

// GuestExportFilters represents filters for guest export
type GuestExportFilters struct {
	Categories []uuid.UUID  `json:"categories"`
	Tags       []uuid.UUID  `json:"tags"`
	RSVPStatus *RSVPStatus  `json:"rsvp_status"`
	Source     *GuestSource `json:"source"`
	Approved   *bool        `json:"approved"`
}

// BulkInvitationRequest represents a request to send bulk invitations
type BulkInvitationRequest struct {
	GuestIDs    []uuid.UUID `json:"guest_ids" binding:"required"`
	Subject     string      `json:"subject" binding:"required"`
	Message     string      `json:"message" binding:"required"`
	ExpiresAt   *time.Time  `json:"expires_at"`
	PrefillData bool        `json:"prefill_data"` // Whether to prefill name/phone from guest data
}

// BulkInvitationResult represents the result of bulk invitation sending
type BulkInvitationResult struct {
	TotalRequested int64                 `json:"total_requested"`
	SuccessCount   int64                 `json:"success_count"`
	FailureCount   int64                 `json:"failure_count"`
	Errors         []BulkInvitationError `json:"errors"`
	Invitations    []Invitation          `json:"invitations"`
}

// BulkInvitationError represents an error during bulk invitation
type BulkInvitationError struct {
	GuestID uuid.UUID `json:"guest_id"`
	Message string    `json:"message"`
}

// BulkRSVPUpdateRequest represents a request to update RSVP status for multiple guests
type BulkRSVPUpdateRequest struct {
	GuestIDs []uuid.UUID `json:"guest_ids" binding:"required"`
	Status   RSVPStatus  `json:"status" binding:"required"`
	Notes    *string     `json:"notes"`
}

// BulkRSVPUpdateResult represents the result of bulk RSVP updates
type BulkRSVPUpdateResult struct {
	TotalRequested int64                 `json:"total_requested"`
	SuccessCount   int64                 `json:"success_count"`
	FailureCount   int64                 `json:"failure_count"`
	Errors         []BulkRSVPUpdateError `json:"errors"`
	UpdatedGuests  []Guest               `json:"updated_guests"`
}

// BulkRSVPUpdateError represents an error during bulk RSVP update
type BulkRSVPUpdateError struct {
	GuestID uuid.UUID `json:"guest_id"`
	Message string    `json:"message"`
}

// BulkDeleteResult represents the result of bulk guest deletion
type BulkDeleteResult struct {
	TotalRequested int64             `json:"total_requested"`
	SuccessCount   int64             `json:"success_count"`
	FailureCount   int64             `json:"failure_count"`
	Errors         []BulkDeleteError `json:"errors"`
}

// BulkDeleteError represents an error during bulk deletion
type BulkDeleteError struct {
	GuestID uuid.UUID `json:"guest_id"`
	Message string    `json:"message"`
}

// CSVImportRow represents a single row from CSV import
type CSVImportRow struct {
	Name         string `csv:"name"`
	Email        string `csv:"email"`
	Phone        string `csv:"phone"`
	Notes        string `csv:"notes"`
	Category     string `csv:"category"`
	Tags         string `csv:"tags"`
	PlusOneName  string `csv:"plus_one_name"`
	PlusOneEmail string `csv:"plus_one_email"`
}

// CSVExportRow represents a single row for CSV export
type CSVExportRow struct {
	Name       string `csv:"name"`
	Email      string `csv:"email"`
	Phone      string `csv:"phone"`
	Notes      string `csv:"notes"`
	RSVPStatus string `csv:"rsvp_status"`
	RSVPDate   string `csv:"rsvp_date"`
	Source     string `csv:"source"`
	Approved   string `csv:"approved"`
	Categories string `csv:"categories"`
	Tags       string `csv:"tags"`
	PlusOnes   string `csv:"plus_ones"`
	SeatInfo   string `csv:"seat_info"`
}

File: frontend/src/types/bulk-operations.ts
/**
 * Bulk operations related types
 */

import { PaginationParams, BaseFilter } from './common';
import { Guest } from './guest';
import { Invitation } from './invitation';

// CSV Import/Export types
export interface GuestExportFilters extends BaseFilter, PaginationParams {
  categories?: string[];
  tags?: string[];
  rsvp_status?: string;
  source?: string;
  approved?: boolean;
}

export interface BulkImportResult {
  total_rows: number;
  successful: number;
  failed: number;
  errors: Array<{
    row: number;
    error: string;
    data?: Record<string, unknown>;
  }>;
  guests: Guest[];
  warnings: Array<{
    row: number;
    warning: string;
    data?: Record<string, unknown>;
  }>;
}

// Bulk invitation types
export interface BulkInvitationRequest {
  emails: string[];
  expires_in_days?: number;
  prefilled_data?: {
    name?: string;
    phone?: string;
    notes?: string;
  };
}

export interface BulkInvitationResult {
  total_requested: number;
  successful: number;
  failed: number;
  errors: Array<{
    email: string;
    error: string;
  }>;
  invitations: Invitation[];
}

// Bulk RSVP update types
export interface BulkRSVPUpdateRequest {
  guest_ids: string[];
  rsvp_status: string;
  notes?: string;
}

export interface BulkRSVPUpdateResult {
  total_requested: number;
  successful: number;
  failed: number;
  errors: Array<{
    guest_id: string;
    error: string;
  }>;
  updated_guests: Guest[];
}

// Bulk delete types
export interface BulkDeleteResult {
  total_requested: number;
  successful: number;
  failed: number;
  errors: Array<{
    guest_id: string;
    error: string;
  }>;
  deleted_guests: string[];
}

// CSV format specification
export interface CSVGuestFormat {
  name: string;
  email: string;
  phone?: string;
  notes?: string;
  dietary_restrictions?: string;
  accessibility_needs?: string;
  emergency_contact?: string;
  emergency_phone?: string;
  plus_one_limit?: number;
  rsvp_status?: string;
  category?: string;
  tags?: string;
}

// Import validation
export interface ImportValidationResult {
  valid: boolean;
  errors: Array<{
    row: number;
    field: string;
    error: string;
    value?: unknown;
  }>;
  warnings: Array<{
    row: number;
    field: string;
    warning: string;
    value?: unknown;
  }>;
  total_rows: number;
  valid_rows: number;
}

// Export options
export interface ExportOptions {
  format: 'csv' | 'xlsx' | 'json';
  include_plus_ones?: boolean;
  include_assignments?: boolean;
  include_communications?: boolean;
  date_range?: {
    start: string;
    end: string;
  };
  fields?: string[];
}

// Bulk operation status
export interface BulkOperationStatus {
  id: string;
  type: 'import' | 'export' | 'invitation' | 'rsvp_update' | 'delete';
  status: 'pending' | 'processing' | 'completed' | 'failed';
  progress: number; // 0-100
  total_items: number;
  processed_items: number;
  errors: string[];
  created_at: string;
  completed_at?: string;
  result?: unknown;
}
